File: packages/snow_draw_engine/lib/draw/elements/types/arrow/arrow_render_primitives.dart
import 'package:meta/meta.dart';

import '../../../types/draw_point.dart';
import '../../../types/element_style.dart';
import 'arrow_core.dart' as core;

/// Endpoint position for arrowhead primitive resolution.
enum ArrowEndpointPosition { start, end }

/// Fill mode attached to a render primitive.
enum ArrowheadPrimitiveFillMode { none, stroke, background }

/// Dash semantics attached to a line primitive.
enum ArrowheadPrimitiveDashMode { solid, inherit, dottedCap }

/// Engine-friendly render primitive descriptor for arrowheads.
@immutable
sealed class ArrowheadRenderPrimitiveData {
  const ArrowheadRenderPrimitiveData();
}

/// Line-segment primitive for arrowhead rendering/hit testing.
@immutable
final class ArrowheadLinePrimitiveData extends ArrowheadRenderPrimitiveData {
  const ArrowheadLinePrimitiveData({
    required this.from,
    required this.to,
    required this.dashMode,
  });

  final DrawPoint from;
  final DrawPoint to;
  final ArrowheadPrimitiveDashMode dashMode;
}

/// Polygon primitive for arrowhead rendering/hit testing.
@immutable
final class ArrowheadPolygonPrimitiveData extends ArrowheadRenderPrimitiveData {
  const ArrowheadPolygonPrimitiveData({
    required this.points,
    required this.fillMode,
  });

  final List<DrawPoint> points;
  final ArrowheadPrimitiveFillMode fillMode;
}

/// Circle primitive for arrowhead rendering/hit testing.
@immutable
final class ArrowheadCirclePrimitiveData extends ArrowheadRenderPrimitiveData {
  const ArrowheadCirclePrimitiveData({
    required this.center,
    required this.radius,
    required this.fillMode,
  });

  final DrawPoint center;
  final double radius;
  final ArrowheadPrimitiveFillMode fillMode;
}

/// Resolves arrowhead primitives strictly via arrow-core.
final class ArrowRenderPrimitives {
  const ArrowRenderPrimitives._();

  /// Builds render primitives for a single endpoint arrowhead.
  static List<ArrowheadRenderPrimitiveData> resolveArrowheadPrimitives({
    required List<DrawPoint> points,
    required ArrowType arrowType,
    required ArrowheadStyle style,
    required StrokeStyle strokeStyle,
    required double strokeWidth,
    required ArrowEndpointPosition position,
    DrawPoint? directionOverride,
  }) {
    // Keep API shape stable while forcing all geometry onto arrow-core.
    final _ = directionOverride;
    if (style == ArrowheadStyle.none || strokeWidth <= 0 || points.length < 2) {
      return const <ArrowheadRenderPrimitiveData>[];
    }

    return _resolveCoreArrowheadPrimitives(
      points: points,
      arrowType: arrowType,
      style: style,
      strokeStyle: strokeStyle,
      strokeWidth: strokeWidth,
      position: position,
    );
  }
}

List<ArrowheadRenderPrimitiveData> _resolveCoreArrowheadPrimitives({
  required List<DrawPoint> points,
  required ArrowType arrowType,
  required ArrowheadStyle style,
  required StrokeStyle strokeStyle,
  required double strokeWidth,
  required ArrowEndpointPosition position,
}) {
  final coreArrowhead = _toCoreRenderableArrowhead(style);
  if (coreArrowhead == null) {
    return const <ArrowheadRenderPrimitiveData>[];
  }

  final curveOps = _buildCoreCurveOps(points: points, arrowType: arrowType);
  if (curveOps.length < 2) {
    return const <ArrowheadRenderPrimitiveData>[];
  }

  final primitives = core.getArrowheadRenderPrimitives(
    core.ArrowheadRenderPrimitivesInput(
      arrowPoints: points
          .map((point) => <double>[point.x, point.y])
          .toList(growable: false),
      strokeWidth: strokeWidth,
      curveOps: curveOps,
      position: _toCoreEndpointPosition(position),
      arrowhead: coreArrowhead,
      strokeStyle: _toCoreStrokeStyle(strokeStyle),
    ),
  );
  if (primitives.isEmpty) {
    return const <ArrowheadRenderPrimitiveData>[];
  }

  final converted = <ArrowheadRenderPrimitiveData>[];
  for (final primitive in primitives) {
    switch (primitive) {
      case core.ArrowheadLinePrimitive():
        converted.add(
          ArrowheadLinePrimitiveData(
            from: DrawPoint(x: primitive.from[0], y: primitive.from[1]),
            to: DrawPoint(x: primitive.to[0], y: primitive.to[1]),
            dashMode: _fromCoreDashMode(primitive.dashMode),
          ),
        );
      case core.ArrowheadPolygonPrimitive():
        converted.add(
          ArrowheadPolygonPrimitiveData(
            points: List<DrawPoint>.unmodifiable(
              primitive.points
                  .map((point) => DrawPoint(x: point[0], y: point[1]))
                  .toList(growable: false),
            ),
            fillMode: _fromCoreFillMode(primitive.fillMode),
          ),
        );
      case core.ArrowheadCirclePrimitive():
        converted.add(
          ArrowheadCirclePrimitiveData(
            center: DrawPoint(x: primitive.center[0], y: primitive.center[1]),
            radius: primitive.diameter / 2,
            fillMode: _fromCoreFillMode(primitive.fillMode),
          ),
        );
    }
  }

  return List<ArrowheadRenderPrimitiveData>.unmodifiable(converted);
}

core.Arrowhead? _toCoreRenderableArrowhead(ArrowheadStyle style) {
  switch (style) {
    case ArrowheadStyle.none:
      return null;
    case ArrowheadStyle.standard:
    case ArrowheadStyle.triangle:
    case ArrowheadStyle.triangleOutline:
    case ArrowheadStyle.square:
    case ArrowheadStyle.dot:
    case ArrowheadStyle.circle:
    case ArrowheadStyle.circleOutline:
    case ArrowheadStyle.diamond:
    case ArrowheadStyle.diamondOutline:
    case ArrowheadStyle.crowfootOne:
    case ArrowheadStyle.crowfootMany:
    case ArrowheadStyle.crowfootOneOrMany:
    case ArrowheadStyle.invertedTriangle:
    case ArrowheadStyle.verticalLine:
      return _toCoreArrowhead(style);
  }
}

core.Arrowhead _toCoreArrowhead(ArrowheadStyle style) {
  switch (style) {
    case ArrowheadStyle.standard:
      return 'arrow';
    case ArrowheadStyle.triangle:
      return 'triangle';
    case ArrowheadStyle.triangleOutline:
      return 'triangle_outline';
    case ArrowheadStyle.square:
      return 'square';
    case ArrowheadStyle.dot:
      return 'dot';
    case ArrowheadStyle.circle:
      return 'circle';
    case ArrowheadStyle.circleOutline:
      return 'circle_outline';
    case ArrowheadStyle.diamond:
      return 'diamond';
    case ArrowheadStyle.diamondOutline:
      return 'diamond_outline';
    case ArrowheadStyle.crowfootOne:
      return 'crowfoot_one';
    case ArrowheadStyle.crowfootMany:
      return 'crowfoot_many';
    case ArrowheadStyle.crowfootOneOrMany:
      return 'crowfoot_one_or_many';
    case ArrowheadStyle.invertedTriangle:
      return 'inverted_triangle';
    case ArrowheadStyle.verticalLine:
      return 'bar';
    case ArrowheadStyle.none:
      return 'arrow';
  }
}

core.ArrowEndpointPosition _toCoreEndpointPosition(
  ArrowEndpointPosition position,
) => position == ArrowEndpointPosition.start
    ? core.arrowEndpointPositionStart
    : core.arrowEndpointPositionEnd;

core.ArrowStrokeStyle _toCoreStrokeStyle(StrokeStyle style) {
  switch (style) {
    case StrokeStyle.solid:
      return 'solid';
    case StrokeStyle.dashed:
      return 'dashed';
    case StrokeStyle.dotted:
      return 'dotted';
  }
}

ArrowheadPrimitiveDashMode _fromCoreDashMode(core.ArrowheadDashMode dashMode) {
  if (dashMode == 'solid') {
    return ArrowheadPrimitiveDashMode.solid;
  }
  if (dashMode == 'dotted-cap') {
    return ArrowheadPrimitiveDashMode.dottedCap;
  }
  if (dashMode == 'inherit') {
    return ArrowheadPrimitiveDashMode.inherit;
  }
  return ArrowheadPrimitiveDashMode.solid;
}

ArrowheadPrimitiveFillMode _fromCoreFillMode(core.ArrowheadFillMode fillMode) {
  if (fillMode == 'stroke') {
    return ArrowheadPrimitiveFillMode.stroke;
  }
  if (fillMode == 'background') {
    return ArrowheadPrimitiveFillMode.background;
  }
  return ArrowheadPrimitiveFillMode.none;
}

List<core.CurvePathOp> _buildCoreCurveOps({
  required List<DrawPoint> points,
  required ArrowType arrowType,
}) {
  if (points.length < 2) {
    return const <core.CurvePathOp>[];
  }

  final ops = <core.CurvePathOp>[
    core.CurvePathOp(
      op: 'move',
      data: <double>[points.first.x, points.first.y],
    ),
  ];
  for (var index = 0; index < points.length - 1; index++) {
    final cubic = arrowType == ArrowType.curved && points.length >= 3
        ? _buildCurvedCubicSegment(points, index)
        : _buildLinearCubicSegment(points[index], points[index + 1]);
    ops.add(
      core.CurvePathOp(
        op: 'bcurveTo',
        data: <double>[
          cubic.control1.x,
          cubic.control1.y,
          cubic.control2.x,
          cubic.control2.y,
          cubic.end.x,
          cubic.end.y,
        ],
      ),
    );
  }
  return List<core.CurvePathOp>.unmodifiable(ops);
}

({DrawPoint control1, DrawPoint control2, DrawPoint end})
_buildLinearCubicSegment(DrawPoint start, DrawPoint end) {
  final control1 = DrawPoint(
    x: start.x + (end.x - start.x) / 3,
    y: start.y + (end.y - start.y) / 3,
  );
  final control2 = DrawPoint(
    x: start.x + (end.x - start.x) * (2 / 3),
    y: start.y + (end.y - start.y) * (2 / 3),
  );
  return (control1: control1, control2: control2, end: end);
}

({DrawPoint control1, DrawPoint control2, DrawPoint end})
_buildCurvedCubicSegment(List<DrawPoint> points, int index) {
  final p0 = index == 0 ? points[index] : points[index - 1];
  final p1 = points[index];
  final p2 = points[index + 1];
  final p3 = index + 2 < points.length ? points[index + 2] : points[index + 1];
  const tension = 1.0;
  final control1 = p1 + (p2 - p0) * (tension / 6);
  final control2 = p2 - (p3 - p1) * (tension / 6);
  return (control1: control1, control2: control2, end: p2);
}

File: packages/snow_draw_rust_engine/src/draw/elements/types/arrow/arrow_render_primitives.rs
#![allow(dead_code)]

use crate::draw::types::draw_point::DrawPoint;
use crate::draw::types::element_style::{ArrowType, ArrowheadStyle, StrokeStyle};

/// Endpoint position for arrowhead primitive resolution.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum ArrowEndpointPosition {
    Start,
    End,
}

/// Fill mode attached to a render primitive.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum ArrowheadPrimitiveFillMode {
    None,
    Stroke,
    Background,
}

/// Dash semantics attached to a line primitive.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum ArrowheadPrimitiveDashMode {
    Solid,
    Inherit,
    DottedCap,
}

/// Engine-friendly render primitive descriptor for arrowheads.
#[derive(Clone, Debug, PartialEq)]
pub enum ArrowheadRenderPrimitiveData {
    Line {
        from: DrawPoint,
        to: DrawPoint,
        dash_mode: ArrowheadPrimitiveDashMode,
    },
    Polygon {
        points: Vec<DrawPoint>,
        fill_mode: ArrowheadPrimitiveFillMode,
    },
    Circle {
        center: DrawPoint,
        radius: f64,
        fill_mode: ArrowheadPrimitiveFillMode,
    },
}

/// Render-ready connector shaft primitives.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct ArrowRenderPrimitives {
    pub shaft_points: Vec<DrawPoint>,
}

impl ArrowRenderPrimitives {
    /// Builds render primitives for a single endpoint arrowhead.
    pub fn resolve_arrowhead_primitives(
        points: &[DrawPoint],
        _arrow_type: ArrowType,
        style: ArrowheadStyle,
        stroke_style: StrokeStyle,
        stroke_width: f64,
        position: ArrowEndpointPosition,
        direction_override: Option<DrawPoint>,
    ) -> Vec<ArrowheadRenderPrimitiveData> {
        if style == ArrowheadStyle::None || stroke_width <= 0.0 || points.len() < 2 {
            return Vec::new();
        }

        let (tip, direction) = match position {
            ArrowEndpointPosition::Start => (
                points[0],
                direction_override.unwrap_or(points[0] - points[1]),
            ),
            ArrowEndpointPosition::End => {
                let last = points.len() - 1;
                (
                    points[last],
                    direction_override.unwrap_or(points[last] - points[last - 1]),
                )
            }
        };

        build_arrowhead_primitives(tip, direction, style, stroke_style, stroke_width)
    }
}

fn build_arrowhead_primitives(
    tip: DrawPoint,
    direction: DrawPoint,
    style: ArrowheadStyle,
    stroke_style: StrokeStyle,
    stroke_width: f64,
) -> Vec<ArrowheadRenderPrimitiveData> {
    let Some(mut dir) = normalize_draw_vector(direction) else {
        return Vec::new();
    };
    let length = arrowhead_length(stroke_width);
    let width = length * 0.6;

    if style == ArrowheadStyle::InvertedTriangle {
        dir = -dir;
    }

    let perp = DrawPoint::new(-dir.y, dir.x);
    let line_dash_mode = match stroke_style {
        StrokeStyle::Dotted => ArrowheadPrimitiveDashMode::DottedCap,
        StrokeStyle::Dashed => ArrowheadPrimitiveDashMode::Inherit,
        StrokeStyle::Solid => ArrowheadPrimitiveDashMode::Solid,
    };

    match style {
        ArrowheadStyle::Standard => {
            let base = subtract_scaled_draw_point(tip, dir, length);
            let left = add_scaled_draw_point(base, perp, width / 2.0);
            let right = add_scaled_draw_point(base, perp, -width / 2.0);
            vec![
                ArrowheadRenderPrimitiveData::Line {
                    from: tip,
                    to: left,
                    dash_mode: line_dash_mode,
                },
                ArrowheadRenderPrimitiveData::Line {
                    from: tip,
                    to: right,
                    dash_mode: line_dash_mode,
                },
            ]
        }
        ArrowheadStyle::Triangle => polygon_primitive(
            triangle_vertices(tip, dir, perp, length, width),
            ArrowheadPrimitiveFillMode::Stroke,
        ),
        ArrowheadStyle::TriangleOutline => polygon_primitive(
            triangle_vertices(tip, dir, perp, length, width),
            ArrowheadPrimitiveFillMode::None,
        ),
        ArrowheadStyle::InvertedTriangle => polygon_primitive(
            triangle_vertices(tip, dir, perp, length, width),
            ArrowheadPrimitiveFillMode::Stroke,
        ),
        ArrowheadStyle::Square => {
            let side = length * 0.6;
            let half = side / 2.0;
            let center = subtract_scaled_draw_point(tip, dir, half);
            let corners = vec![
                add_draw_points(
                    add_scaled_draw_point(center, perp, half),
                    add_scaled_draw_point(DrawPoint::ZERO, dir, half),
                ),
                add_draw_points(
                    add_scaled_draw_point(center, perp, -half),
                    add_scaled_draw_point(DrawPoint::ZERO, dir, half),
                ),
                add_draw_points(
                    add_scaled_draw_point(center, perp, -half),
                    add_scaled_draw_point(DrawPoint::ZERO, dir, -half),
                ),
                add_draw_points(
                    add_scaled_draw_point(center, perp, half),
                    add_scaled_draw_point(DrawPoint::ZERO, dir, -half),
                ),
            ];
            polygon_primitive(corners, ArrowheadPrimitiveFillMode::Stroke)
        }
        ArrowheadStyle::Dot => vec![ArrowheadRenderPrimitiveData::Circle {
            center: subtract_scaled_draw_point(tip, dir, length * 0.3),
            radius: length * 0.3,
            fill_mode: ArrowheadPrimitiveFillMode::Stroke,
        }],
        ArrowheadStyle::Circle => vec![ArrowheadRenderPrimitiveData::Circle {
            center: subtract_scaled_draw_point(tip, dir, length * 0.3),
            radius: length * 0.3,
            fill_mode: ArrowheadPrimitiveFillMode::Stroke,
        }],
        ArrowheadStyle::CircleOutline => vec![ArrowheadRenderPrimitiveData::Circle {
            center: subtract_scaled_draw_point(tip, dir, length * 0.3),
            radius: length * 0.3,
            fill_mode: ArrowheadPrimitiveFillMode::None,
        }],
        ArrowheadStyle::Diamond => {
            let base = subtract_scaled_draw_point(tip, dir, length);
            let mid = subtract_scaled_draw_point(tip, dir, length / 2.0);
            let left = add_scaled_draw_point(mid, perp, width / 2.0);
            let right = add_scaled_draw_point(mid, perp, -width / 2.0);
            polygon_primitive(
                vec![tip, left, base, right],
                ArrowheadPrimitiveFillMode::Stroke,
            )
        }
        ArrowheadStyle::DiamondOutline => {
            let base = subtract_scaled_draw_point(tip, dir, length);
            let mid = subtract_scaled_draw_point(tip, dir, length / 2.0);
            let left = add_scaled_draw_point(mid, perp, width / 2.0);
            let right = add_scaled_draw_point(mid, perp, -width / 2.0);
            polygon_primitive(
                vec![tip, left, base, right],
                ArrowheadPrimitiveFillMode::None,
            )
        }
        ArrowheadStyle::CrowfootOne => {
            let base = subtract_scaled_draw_point(tip, dir, length);
            let left = add_scaled_draw_point(base, perp, width / 2.0);
            let right = add_scaled_draw_point(base, perp, -width / 2.0);
            vec![ArrowheadRenderPrimitiveData::Line {
                from: left,
                to: right,
                dash_mode: line_dash_mode,
            }]
        }
        ArrowheadStyle::CrowfootMany => {
            let base = subtract_scaled_draw_point(tip, dir, length);
            let left = add_scaled_draw_point(base, perp, width / 2.0);
            let right = add_scaled_draw_point(base, perp, -width / 2.0);
            vec![
                ArrowheadRenderPrimitiveData::Line {
                    from: tip,
                    to: left,
                    dash_mode: line_dash_mode,
                },
                ArrowheadRenderPrimitiveData::Line {
                    from: tip,
                    to: right,
                    dash_mode: line_dash_mode,
                },
            ]
        }
        ArrowheadStyle::CrowfootOneOrMany => {
            let base = subtract_scaled_draw_point(tip, dir, length);
            let left = add_scaled_draw_point(base, perp, width / 2.0);
            let right = add_scaled_draw_point(base, perp, -width / 2.0);
            vec![
                ArrowheadRenderPrimitiveData::Line {
                    from: tip,
                    to: left,
                    dash_mode: line_dash_mode,
                },
                ArrowheadRenderPrimitiveData::Line {
                    from: tip,
                    to: right,
                    dash_mode: line_dash_mode,
                },
                ArrowheadRenderPrimitiveData::Line {
                    from: left,
                    to: right,
                    dash_mode: line_dash_mode,
                },
            ]
        }
        ArrowheadStyle::VerticalLine => {
            let half = width / 2.0;
            let left = add_scaled_draw_point(tip, perp, half);
            let right = add_scaled_draw_point(tip, perp, -half);
            vec![ArrowheadRenderPrimitiveData::Line {
                from: left,
                to: right,
                dash_mode: line_dash_mode,
            }]
        }
        ArrowheadStyle::None => Vec::new(),
    }
}

fn triangle_vertices(
    tip: DrawPoint,
    dir: DrawPoint,
    perp: DrawPoint,
    length: f64,
    width: f64,
) -> Vec<DrawPoint> {
    let base = subtract_scaled_draw_point(tip, dir, length);
    let left = add_scaled_draw_point(base, perp, width / 2.0);
    let right = add_scaled_draw_point(base, perp, -width / 2.0);
    vec![tip, left, right]
}

fn polygon_primitive(
    points: Vec<DrawPoint>,
    fill_mode: ArrowheadPrimitiveFillMode,
) -> Vec<ArrowheadRenderPrimitiveData> {
    vec![ArrowheadRenderPrimitiveData::Polygon { points, fill_mode }]
}

fn arrowhead_length(stroke_width: f64) -> f64 {
    stroke_width * 4.0 + 12.0
}

fn add_draw_points(a: DrawPoint, b: DrawPoint) -> DrawPoint {
    DrawPoint::new(a.x + b.x, a.y + b.y)
}

fn add_scaled_draw_point(a: DrawPoint, direction: DrawPoint, scale: f64) -> DrawPoint {
    DrawPoint::new(a.x + direction.x * scale, a.y + direction.y * scale)
}

fn subtract_scaled_draw_point(a: DrawPoint, direction: DrawPoint, scale: f64) -> DrawPoint {
    DrawPoint::new(a.x - direction.x * scale, a.y - direction.y * scale)
}

fn normalize_draw_vector(value: DrawPoint) -> Option<DrawPoint> {
    let length = (value.x * value.x + value.y * value.y).sqrt();
    if length == 0.0 {
        return None;
    }
    Some(DrawPoint::new(value.x / length, value.y / length))
}
